[NVIDIA TF] Optimize embedding_lookup_sparse using new grad op [PART 3/3]

benbarsdell opened the pull request:
Sequel to https://github.com/tensorflow/tensorflow/pull/61294

(This is a rebased version of the approved https://github.com/benbarsdell/tensorflow/pull/2)

cc @bfontain @nluehr @pjannaty

benbarsdell committed:
Add sparse_gradient to sparse segment reduce ops

- Adds a sparse_gradient option to the sparse segment reduce ops. When
  this is enabled, the V2 gradient op is used and the gradient is
  returned as an IndexedSlices object instead of a dense tensor.

benbarsdell committed:
Skip unnecessary ops in embedding_lookup_sparse

- Skips the unique + embedding_lookup calls when possible and instead
  passes the embedding parameters directly to the sparse segment
  reduction op, which is significantly more efficient.
- Also adds a benchmark that demonstrates a 2x speedup from this change
  (on a Titan V GPU).

cantonios reviewed (changes requested):
Can you add a quick note under Release.md about the new op/API attributes?

benbarsdell committed:
Note segment reduce op/API changes in Release.md

benbarsdell commented:
Thanks, I've added notes to Release.md.

cantonios commented on tensorflow/python/ops/math_grad.py:
```diff
@@ -19,6 +19,7 @@
 from tensorflow.python.eager import context
 from tensorflow.python.framework import constant_op
 from tensorflow.python.framework import dtypes
+from tensorflow.python.framework import indexed_slices as indexed_slices_lib
```
We need to add this dependency to the BUILD file for math_grad

benbarsdell replied:
Done

cantonios commented on tensorflow/python/ops/embedding_ops.py:
```diff
@@ -14,6 +14,7 @@
 # ==============================================================================
 """Operations for embeddings."""
 
+from tensorflow.python.compat import compat
```
We need to add this dependency to the BUILD file for embedding_ops

benbarsdell replied:
Done

cantonios replied:
Thanks

benbarsdell committed:
Add two missing build dependencies

cantonios commented on tensorflow/python/ops/embedding_ops.py:
```diff
@@ -1007,9 +1008,26 @@ def embedding_lookup_sparse_impl(
+  segment_ids = ops.convert_to_tensor(segment_ids, name="segment_ids")
+  if (len(params) == 1 and
+      not isinstance(params[0], resource_variable_ops.BaseResourceVariable)):
+    params = [ops.convert_to_tensor(params[0], name="params")]
```
We have some tests failing because of this line.
```
TypeError: Converting ShardedVariable to tensor in embedding lookup ops is disallowed.
```
For example, `//tensorflow/python/distribute:sharded_variable_test`

cantonios replied:
Ack, will pull in and check on our end, thanks!

benbarsdell committed:
Fix ShardedVariable bug in embedding_lookup_sparse

- Fixes `TypeError: Converting ShardedVariable to tensor in embedding
  lookup ops is disallowed.`

cantonios commented on tensorflow/python/ops/BUILD:
```diff
@@ -1561,6 +1561,8 @@ py_strict_library(
         ":sparse_ops",
         ":variables",
+        "//tensorflow/python/compat",
+        "//tensorflow/python/distribute:sharded_variable",
```
Unfortunately this creates a circular dependency
```
embedding_ops --> shared_variable --> embedding_ops
```

cantonios commented on tensorflow/python/ops/embedding_ops.py:
```diff
@@ -1007,9 +1009,27 @@ def embedding_lookup_sparse_impl(
+  # Ensure we can query the devices below.
+  segment_ids = ops.convert_to_tensor(segment_ids, name="segment_ids")
+  if (len(params) == 1 and
+      not isinstance(params[0], (resource_variable_ops.BaseResourceVariable,
```
Could you use `core.Tensor` and `composite_tensor.CompositeTensor`?  That should cover these two cases, and more.

Or, there is a `tensor_util.is_tf_type(x)` function that covers more types.

benbarsdell replied:
Thanks, `is_tf_type` didn't work but the first suggestion did. I got the test to pass now. (Also rebased to fix merge conflict).

cantonios commented on tensorflow/python/kernel_tests/nn_ops/embedding_ops_test.py:
```diff
@@ -909,6 +913,86 @@ def test_incompatible_types(self):
+        x = resource_variable_ops.ResourceVariable(x)
+
+        def forward(x_):
+          y_ = embedding_ops.embedding_lookup_sparse(
```
I'm getting a failure on CPU tests around here:
```
bazel test //tensorflow/python/kernel_tests/nn_ops:embedding_ops_test_cpu --compilation_mode=opt --copt=-UNDEBUG
```
```
Detected at node 'embedding_lookup_sparse/embedding_lookup' defined at (most recent call last):
    ...
    File "tensorflow/python/kernel_tests/nn_ops/embedding_ops_test.py", line 978, in testResourceVariableGradientEmbeddingLookupSparse
    File "tensorflow/python/kernel_tests/nn_ops/embedding_ops_test.py", line 970, in forward

Node: 'embedding_lookup_sparse/embedding_lookup'
Container localhost does not exist. (Could not find resource: localhost/Variable)
	 [[{{node embedding_lookup_sparse/embedding_lookup}}]]
```

benbarsdell replied:
I could only reproduce this by running under `@test_util.run_deprecated_v1`, but I've added the required variable initialization line to fix it.

benbarsdell committed:
Fix embedding ops test for TF1 mode

benbarsdell committed:
Add missing import deps for embedding lookup tests

benbarsdell committed:
Bump compat horizon for new embedding ops

benbarsdell commented on tensorflow/python/eager/benchmarks_test.py:
```diff
@@ -33,6 +33,7 @@
 import numpy as np
 
 from tensorflow.python import pywrap_tfe
+from tensorflow.python.compat import compat as forward_compat
```
I couldn't work out how to reproduce these errors, but I've added the deps that should fix them.

benbarsdell commented on tensorflow/python/kernel_tests/math_ops/segment_reduction_ops_test.py:
```diff
@@ -23,10 +23,12 @@
 from tensorflow.python.framework import constant_op
 from tensorflow.python.framework import dtypes as dtypes_lib
 from tensorflow.python.framework import errors_impl
+from tensorflow.python.framework import indexed_slices
```
Deps added.

benbarsdell commented on tensorflow/python/kernel_tests/nn_ops/embedding_ops_test.py:
```diff
@@ -20,6 +20,7 @@
 
 import numpy as np
 
+from tensorflow.python.compat import compat as forward_compat
```
Deps added.

benbarsdell commented:
I've also rebased and bumped the compat horizon.

benbarsdell commented:
@cantonios are there still some internal tests failing?
